core/common/scale/advanced.hpp

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_SCALE_ADVANCED_HPP
+#define KAGOME_SCALE_ADVANCED_HPP
+
+#include "common/scale/compact.hpp"
+#include "common/scale/fixedwidth.hpp"
+#include "common/scale/type_decoder.hpp"
+#include "common/scale/type_encoder.hpp"
+#include "util.hpp"
+
+namespace kagome::common::scale::advanced {
+  /**
+   *  @brief encodeCollection function scale-encodes
+   *         collection of same type values
+   *  @tparam T Item type stored in collection
+   *  @param collection source vector items
+   *  @return byte array containing encoded collection
+   */
+  template <class T>
+  EncodeResult encodeCollection(const std::vector<T> &collection) {
+    auto header = compact::encodeInteger(BigInteger{collection.size()});
+    if (!isSucceeded(header)) {
+      return EncodeResult::ErrorType{EncodeError::kEncodeHeaderError};
+    }
+
+    ByteArray result;
+    header.match(
+        [&collection, &result](const expected::Value<ByteArray> &v) {
+          if (std::is_integral<T>()) {
+            size_t bytes_required = v.value.size();
+            bytes_required += collection.size() * sizeof(T);
+            result.reserve(bytes_required);
+          }
```

Comment:
integral types are simple built-in types like char, wchar_t, all integers, I mean the types which size can be calculated beforehand, and this fits most of situations we are going to have in polkadot. In order to get rid of unnecessary relocations, which take a lot of time I calculate and reserve space. If type is not integral calculating size like this can be useless and wrong.